Docs: Add important note on using `iterall` and `iterdict` (#6126)
aiidateam/aiida-core: docs/source/howto/query.rst

    all_results_l = qb.all()            # Returns a list of lists

.. tip::

>     If your query only has a single projection, you can use ``flat=True`` in the ``first`` and ``all`` methods.
>     This will return just the result value or a flat list, respectively, and no longer has the unnecessary internal list.

Author: ```suggestion
    If your query only has a single projection, use ``flat=True`` in the ``first`` and ``all`` methods to return a single value or a flat list, respectively.
```

    for entry in qb.iterall():
        # do something with a single entry in the query result

.. important::

    When the query result is expected to be large, do not use ``dict`` and ``all`` but use the generator variants ``iterdict`` and ``iterall``.
    The ``dict`` and ``all`` methods load the entire query result in memory which can lead to out-of-memory errors.
    It also makes updating the data very slow as each change results in a single database commit.

>     The ``iterdict`` and ``iterall`` generators open a database transaction, committing all changes only once at the end of the generator.

Reviewer: Maybe we should add a note that in case e.g. a loop running over the generator gets interrupted, none of the changes are actually committed, or stress this somehow.

>     When the query result is expected to be large, do not use ``dict`` and ``all`` but use the generator variants ``iterdict`` and ``iterall``.
>     The ``dict`` and ``all`` methods load the entire query result in memory which can lead to out-of-memory errors.
>     It also makes updating the data very slow as each change results in a single database commit.
>     The ``iterdict`` and ``iterall`` generators open a database transaction, committing all changes only once at the end of the generator.

Reviewer 2: Here my suggestion
```suggestion
    When looping over the result of a query, use the ``iterall`` (or ``iterdict``) generators instead of ``all`` (or ``dict``).
    This avoids loading the entire query result into memory, and it also delays committing changes made to AiiDA objects inside the loop until the end of the loop is reached.
```
